### ISSUE TITLE
Data::Dumper ignores ^M in scalars?

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8959]
Migrated from [rt.perl.org#43617](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=43617) (status was 'resolved')

Searchable as RT43617$

**Comment by p5pRT** [comment]
### From hag@linnaean.org

#### Created by hag@linnaean.org



Data&#8203;::Dumper appears to ignore ^M in scalar variables on this system;
this does not strike me as a feature\.

e\.g\.&#8203;:

perl \-e 'use Data&#8203;::Dumper; print Dumper\(chr\(13\)\)'

produces&#8203;:

$VAR1 = '';

It will do this with or without useqq set\.  Other values behave as expected\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=high

Site configuration information for perl v5.8.8:

Configured by root at Thu Jun  7 12:09:46 EDT 2007.

Summary of my perl5 (revision 5 version 8 subversion 8) configuration:
  Platform:
    osname=netbsd, osvers=3.1, archname=i386-netbsd-thread-multi
    uname='netbsd fast-wide.linnaean.org 3.1 netbsd 3.1 (fast_wide) #0: thu jun 7 03:23:45 edt 2007 root@fast-wide.linnaean.org:usrsrcsysarchi386compilefast_wide i386 '
    config_args='-sde -Darchname=i386-netbsd -Dcc=cc -Doptimize=-O2 -pthread -I/usr/include -Duseshrplib -Ui_malloc -Uusemymalloc -Uinstallusrbinperl -Dinstallstyle=lib/perl5 -Dprefix=/usr/pkg -Dsiteprefix=/usr/pkg -Dvendorprefix=/usr/pkg -Dscriptdir=/usr/pkg/lib/perl5/bin -Dsitescript=/usr/pkg/lib/perl5/site_perl/bin -Dvendorscript=/usr/pkg/lib/perl5/vendor_perl/bin -Dsitebin=/usr/pkg/lib/perl5/site_perl/bin -Dvendorbin=/usr/pkg/lib/perl5/vendor_perl/bin -Dprivlib=/usr/pkg/lib/perl5/5.8.0 -Dsitelib=/usr/pkg/lib/perl5/site_perl/5.8.0 -Dvendorlib=/usr/pkg/lib/perl5/vendor_perl/5.8.0 -Dsitelib_stem=/usr/pkg/lib/perl5/site_perl -Dvendorlib_stem=/usr/pkg/lib/perl5/vendor_perl -Dman1ext=1 -Dman1dir=/usr/pkg/lib/perl5/man/man1 -Dsiteman1dir=/usr/pkg/lib/perl5/site_perl/man/man1 -Dvendorman1dir=/usr/pkg/lib/perl5/vendor_perl/man/man1 -Dman3ext=3 -Dman3dir=/usr/pkg/lib/perl5/man/man3 -Dsiteman3dir=/usr/pkg/lib/perl5/site_perl/man/man3 -Dvendorman3dir=/usr/pkg/lib/perl5/vendor_perl/ma
 n/man3 -Daphostname=/bin/hostname -Dln=/bin/ln -Dsed=/usr/bin/sed -Dsh=/bin/sh -Dissymlink=test -h -Duseithreads -Dlibswanted=m crypt '
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -I/usr/pkg/include',
    optimize='-O2 -pthread -I/usr/include',
    cppflags='-fno-strict-aliasing -pipe -I/usr/pkg/include'
    ccversion='', gccversion='3.3.3 (NetBSD nb3 20040520)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-R/usr/pkg/lib  -L/usr/pkg/lib'
    libpth=/usr/lib /usr/pkg/lib
    libs=-lm -lcrypt -lpthread
    perllibs=-lm -lcrypt -lpthread
    libc=/lib/libc.so, so=so, useshrplib=true, libperl=libperl.so
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E  -Wl,-R/usr/pkg/lib/perl5/5.8.0/i386-netbsd-thread-multi/CORE'
    cccdlflags='-DPIC -fPIC ', lddlflags='-Wl,-R/usr/pkg/lib --whole-archive -shared  -L/usr/pkg/lib'

Locally applied patches:
    


@INC for perl v5.8.8:
    /usr/pkg/lib/perl5/site_perl/5.8.0/i386-netbsd-thread-multi
    /usr/pkg/lib/perl5/site_perl/5.8.0
    /usr/pkg/lib/perl5/site_perl
    /usr/pkg/lib/perl5/vendor_perl/5.8.0/i386-netbsd-thread-multi
    /usr/pkg/lib/perl5/vendor_perl/5.8.0
    /usr/pkg/lib/perl5/vendor_perl
    /usr/pkg/lib/perl5/5.8.0/i386-netbsd-thread-multi
    /usr/pkg/lib/perl5/5.8.0
    .


Environment for perl v5.8.8:
    HOME=/home/hag
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/hag/etc/init/common/bin:/home/hag/bin/share:/usr/pkg/bin:/usr/pkg/sbin:/usr/local/gnubin:/usr/gnu/bin:/usr/local/bin:/usr/X11R6/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/local/etc:/etc:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/pkg/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

On 7/7/07\, via RT Daniel Hagerty \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Daniel Hagerty
> \# Please include the string&#8203;:  \[perl \#43617\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=43617 >
>
>
> This is a bug report for perl from hag@&#8203;linnaean\.org\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.8\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
>
> Data&#8203;::Dumper appears to ignore ^M in scalar variables on this system;
> this does not strike me as a feature\.
>
> e\.g\.&#8203;:
>
> perl \-e 'use Data&#8203;::Dumper; print Dumper\(chr\(13\)\)'
>
> produces&#8203;:
>
> $VAR1 = '';
>
> It will do this with or without useqq set\.  Other values behave as expected\.

Can you please tell us what Dumper version you are using by doing

perl \-MData&#8203;::Dumper=1000 \-e1

When i do the code you have outlined i get&#8203;:

';AR1 = '

and if i enable Useqq i get&#8203;:

$VAR1 = "\\r";

Which shows its working properly in version 2\.121\_02

cheers\,
yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

I'm marking this as resolved \+ not a bug based on&#8203;:

a\) not getting new feedback from the submitter
b\) not being able to reproduce this \(with perl554\-perl5100\)
c\) odd perl installation\. \(Installing 5\.8\.8 but using lib dir of 5\.8\.0 
\- /usr/pkg/lib/perl5/5\.8\.0\)

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From hag@linnaean.org

&nbsp;    Apologies on not having responded earlier\, I probably missed the
response\.

&nbsp;    FWIW\, useqq now produces expected results; usage without useqq
continues to produce an empty string\.  Tried this with perl 5\.8\.8 w/
Data&#8203;::Dumper 2\.121\_08 on both netbsd 4\.0 and debian 4\.0 systems\.

&nbsp;    If one is to believe that Data&#8203;::Dumper's intent is to produce a
string that can be eval'd to produce a copy of the original object
like it's manual page indicates\, this appears to be a bug\.

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

On Mon\, 28 Apr 2008 00&#8203;:18&#8203;:28 \+0200\, Daniel Hagerty \<hag@&#8203;linnaean\.org>  
wrote&#8203;:

>     Apologies on not having responded earlier\, I probably missed the
> response\.

Thank you for replying\.

>     FWIW\, useqq now produces expected results; usage without useqq
> continues to produce an empty string\.

\#\!/usr/bin/perl

use Data&#8203;::Dumper;
print Dumper\("a\\rb"\);


Does it produce an empty string or do you see \- in the shell \- something  
like&#8203;:  b';R1 = 'a'
Or using | od \-tx1&#8203;: 0000000 24 56 41 52 31 20 3d 20 27 61 0d 62 27 3b 0a
\(a = 61\, \\r = 0d\, b = 62\)


> Tried this with perl 5\.8\.8 w/
> Data&#8203;::Dumper 2\.121\_08 on both netbsd 4\.0 and debian 4\.0 systems\.
>
>     If one is to believe that Data&#8203;::Dumper's intent is to produce a
> string that can be eval'd to produce a copy of the original object
> like it's manual page indicates\, this appears to be a bug\.

Actually\, it can be eval'ed without a problem\.
You can even write it in a file\, read that file and eval it\.

What you can not do is copy the output from the shell/console\, paste it in  
another file and eval that\.
The handling of  ^M \(or \\r\) is seen by the shell \(or atleast I think\) as a  
way how to control the display of the output\.
The same is true for ANSI escapes\. \(if your  
shel/terminal/console/display/\.\.\. supports them\)

This is because the shell/console reads them and changes the text that is  
displayed\.
If you do not want the shell to do that then you can set  $Useqq = 1   
which prevents them from being outputted\.

In all\, I do not think this is a bug\.


Try for example&#8203;:

\#\!/usr/bin/perl

use Data&#8203;::Dumper;
my $foo = Dumper\("a\\rb"\);
print "Foo = $foo";
eval $foo;

$Data&#8203;::Dumper&#8203;::Useqq = 1;
print "eval \\$foo = " \. Dumper\($VAR1\);
$VAR1 = "";

my $fn = "/tmp/foo\_$$\_" \. time;
open my $fh\, ">"\, $fn or die "Can't open file $fn&#8203;: $\!";
print $fh $foo;
close $fh;

require $fn;
unlink $fn;
print "require $fn = " \. Dumper\($VAR1\);
$VAR1 = "";

\_\_END\_\_

I hope you now agree with me that it is not a bug\.

Kind regards\,

Bram

**Comment by p5pRT** [comment]
### From p5p@spam.wizbit.be

Off list reply from Daniel Hagerty&#8203;:

Date&#8203;:     Mon\, 28 Apr 2008 05&#8203;:46&#8203;:15 \-0400 \[11&#8203;:46&#8203;:15 CEST\]    
From&#8203;:     Daniel Hagerty \<hag@&#8203;linnaean\.org>    
Subject&#8203;:     Re&#8203;: \[perl \#43617\] \[RESOLVED\] Data&#8203;::Dumper ignores ^M in 
scalars?    

&nbsp;    Well\, you have me here\.  I'm a little surprised\, as emacs as a
terminal normally leaves things more or less untouched \(e\.g\. ascii NUL
will print\)\, but presumably it's treating ^M specially\.  Must track
that down and fix it\.  BTW\, cat \-v is one of the more canonical ways
of undoing "the terminal is hiding something from you" compared to
od/hd\.

&nbsp;    I know I lost a couple of hours on this originally where the ^M
really wasn't there\, and the fact that useqq's behavior changed
between then and now suggests that \*somebody\* has been here\.  In any
event\, I agree\, no bug at this time\.
